Backend/Remora.Discord.Voice/DiscordVoiceClient.cs

```diff
@@ -0,0 +1,1258 @@
+            IAudioTranscoderService? transcoder = null,
+            CancellationToken ct = default
+        )
+        {
+            if (this.ConnectionStatus is not GatewayConnectionStatus.Connected || !_dataService.IsConnected)
+            {
+                return new InvalidOperationError("Gateway is not connected.");
+            }
+
+            if (speakingFlags is not SpeakingFlags.Microphone or SpeakingFlags.Priority)
+            {
+                return new InvalidOperationError($"Speaking flags must not be set to {nameof(SpeakingFlags.Microphone)} or {nameof(SpeakingFlags.Priority)}.");
+            }
+
+            if (transcoder is null)
+            {
+                transcoder = _defaultTranscoder;
+                await transcoder.ResetAsync(ct).ConfigureAwait(false);
+            }
+
+            if (!transcoder.IsInitialized)
+            {
+                return new InvalidOperationError("The provided transcoder must first be initialized.");
+            }
+
+            // Take a local copy of the SSRC so we can still send a voice state update when the client is stopped
+            var ssrc = _voiceServerConnectionDetails!.SSRC;
+            var needsRelease = false;
+
+            var sampleSize = transcoder.SampleSize;
+            var inputBuffer = MemoryPool<byte>.Shared.Rent(sampleSize);
+            var opusBuffer = MemoryPool<byte>.Shared.Rent(sampleSize);
+
+            try
+            {
+                if (!await _transmitSemaphore.WaitAsync(10, ct).ConfigureAwait(false))
+                {
+                    return new InvalidOperationError("This client is already transmitting audio.");
+                }
+
+                needsRelease = true;
+
+                // From https://github.com/DSharpPlus/DSharpPlus/blob/master/DSharpPlus.VoiceNext/VoiceNextConnection.cs
+                var synchronizerTicks = (double)Stopwatch.GetTimestamp();
+                var synchronizerResolution = Stopwatch.Frequency * 0.005;
+                var tickResolution = 10_000_000.0 / Stopwatch.Frequency;
+
+                EnqueueCommand
+                (
+                    new VoiceSpeakingCommand
+                    (
+                        speakingFlags,
+                        0,
+                        ssrc
+                    )
+                );
+
+                while (!ct.IsCancellationRequested)
+                {
+                    var read = await audioStream.ReadAsync
+                    (
+                        inputBuffer.Memory[0..sampleSize],
+                        ct
+                    ).ConfigureAwait(false);
+
+                    if (read != sampleSize)
+                    {
+                        break; // TODO: Does this cut off audio in some cases?
+                    }
+
+                    var encodeResult = await transcoder.EncodeAsync
+                    (
+                        inputBuffer.Memory[..read],
+                        opusBuffer.Memory[..read],
```

Comment:
In addendum to the above, won't this throw an `IndexOutOfBoundsException`? You'd be off by 1, I think.

On top of this, wouldn't this be the same as the sample size anyway, since you `break` from the loop if it's not the same?

Reply:
Well, it hasn't yet, which I take as a good sign 😛. The end value is exclusive.

Yes, you're absolutely correct that it's the sample size. I could use either or, but if for whatever reason down the line we decide to transcode just the data that's been read regardless of sample size, it saves a whopping two seconds of code editing 😂.